File: Experimental/2023/TrainingBot/src/main/cpp/commands/TankDrive.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "commands/TankDrive.h"

TankDrive::TankDrive(DriveBase* driveBase, std::function<double()> leftSpeed, std::function<double()> rightSpeed) {
  m_driveBase = driveBase;
  m_leftSpeed = leftSpeed;
  m_rightSpeed = rightSpeed;

  // Tell the rest of the code that we use this when we're running.
  AddRequirements(m_driveBase);
}

// Called when the command is initially scheduled.
void TankDrive::Initialize() {
  m_driveBase->TankDrive(m_leftSpeed(), m_rightSpeed());
}

// Called repeatedly when this Command is scheduled to run
void TankDrive::Execute() {
  m_driveBase->TankDrive(m_leftSpeed(), m_rightSpeed());
}

// Called once the command ends or is interrupted.
void TankDrive::End(bool interrupted) {
  m_driveBase->Stop();
}

// Returns true when the command should end.
bool TankDrive::IsFinished() {
  return false;
}

File: Quasics2026Code/CrescendoJava-2026/src/main/java/frc/robot/commands/TankDrive.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package frc.robot.commands;
import edu.wpi.first.wpilibj2.command.Command;
import frc.robot.subsystems.Drivebase;
import java.util.function.Supplier;

public class TankDrive extends Command {
  private final Drivebase m_drivebase;
  private final Supplier<Double> m_leftSupplier;
  private final Supplier<Double> m_rightSupplier;

  /** Creates a new TankDrive. */
  public TankDrive(Drivebase drivebase, Supplier<Double> leftSupplier,
      Supplier<Double> rightSupplier) {
    m_leftSupplier = leftSupplier;
    m_rightSupplier = rightSupplier;
    m_drivebase = drivebase;

    addRequirements(drivebase);
  }

  // Called when the command is initially scheduled.
  @Override
  public void initialize() {
    updateSpeeds();
  }

  // Called every time the scheduler runs while the command is scheduled.
  @Override
  public void execute() {
    updateSpeeds();
  }

  private void updateSpeeds() {
    final double leftInput = m_leftSupplier.get();
    final double rightInput = m_rightSupplier.get();

    double leftSpeed = leftInput;
    double rightSpeed = rightInput;

    m_drivebase.setSpeeds(leftSpeed, rightSpeed);
  }

  // Called once the command ends or is interrupted.
  @Override
  public void end(boolean interrupted) {
    m_drivebase.stop();
  }
}
